src/driver/shareshift_test_point_generator.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2022 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Generates test points for the ShareShift use case."""
+
+import numpy as np
+from typing import Iterable, List
+from wfa_planning_evaluation_framework.data_generators.data_set import DataSet
+from wfa_planning_evaluation_framework.driver.test_point_generator import (
+    TestPointGenerator,
+)
+
+
+class ShareShiftTestPointGenerator(TestPointGenerator):
+    """Generates test points that proportionally shift budget to or from a publisher."""
+
+    def __init__(
+        self,
+        dataset: DataSet,
+        campaign_spend_fractions: np.ndarray,
+        shift_fraction_choices: List[float] = [-1.0, -0.5, 0.5, 1.0],
+    ):
+        """Constructs a ShareShiftTestPointGenerator.
+
+        Suppose that the original spend vector is [s_1, ..., s_p].
+        This class generates p * m testing points v with:
+            v[i] = s_i + s_i * a_k and
+            v[j] = s_j - s_j / (sum(s) - s[i]) * (s_i * a_k) for j != i,
+        iterating over 1 <= i <= p and k = 1, ..., m.  Here, a_1, ..., a_m are a list
+        of shift fractions.
+        (Note that a testing point will be dropped if any of its element is negative,
+        so there could be fewer than p * m testing points.)
+
+        For example, if
+            * the original spend is [1, 2, 3]
+            * we consider shift fraction = - 0.5, 0.5,
+        then this class generates 2 * 3 testing points with the following spend
+        vectors:
+            * [1 - 0.5 * 1, 2 + (0.5 * 1) * 2 / (2 + 3), 3 + (0.5 * 1) * 3 / (2 + 3)]
+            * [1 + 0.5 * 1, 2 - (0.5 * 1) * 2 / (2 + 3), 3 - (0.5 * 1) * 3 / (2 + 3)]
+            * [1 + (0.5 * 2) * 1 / (1 + 3), 2 - 0.5 * 2, 3 + (0.5 * 2) * 3 / (1 + 3)]
+            * [1 - (0.5 * 2) * 1 / (1 + 3), 2 + 0.5 * 2, 3 - (0.5 * 2) * 3 / (1 + 3)]
+            * [1 + (0.5 * 3) * 1 / (1 + 2), 2 + (0.5 * 3) * 2 / (1 + 2), 3 - 3 * 0.5]
+            * [1 - (0.5 * 3) * 1 / (1 + 2), 2 - (0.5 * 3) * 2 / (1 + 2), 3 + 3 * 0.5]
+
+        Args:
+            dataset:  The DataSet for which test points are to be generated.
+            campaign_spend_fractions:  The campaign_spend / inventory_spend
+                at each publisher, where each inventory_spend is given by
+                dataset._max_spends.
```

Comment:
Similar to what Rieman recently suggested, probably add the dimension of each input argument?